### PR TITLE
Refactor: grouped TS splitter into helper function

### PR DESCRIPTION
**Pull request by paxcema**
## Why

To easily access this method as a helper function from e.g. the forecast visualizer

## How

Simply abstracting the logic into its own function.

### Affected modules

`lightwood/data/splitter`

### CODE REVIEW

File: lightwood/data/splitter.py [hunk 1]
```diff
@@ -7,6 +7,9 @@
 
 
 def splitter(data: pd.DataFrame, k: int, tss: TimeseriesSettings) -> List[pd.DataFrame]:
+    """
+    Splits a dataframe into k equally-sized folds.
```
Comment (hakunanatasha) [inline lightwood/data/splitter.py:11]: I hate the term folds here as it's not clear what "folds" should be.

Are folds equally sized to the training data? are folds subsets of the data? Please clarify.

File: lightwood/data/splitter.py [hunk 2]
```diff
@@ -19,16 +22,27 @@ def splitter(data: pd.DataFrame, k: int, tss: TimeseriesSettings) -> List[pd.Dat
             folds = np.array_split(data, k)
         else:
             gcols = tss.group_by
-            all_group_combinations = list(product(*[data[gcol].unique() for gcol in tss.group_by]))
-            folds = [pd.DataFrame() for _ in range(k)]
-            for group in all_group_combinations:
-                subframe = data
-                for idx, gcol in enumerate(gcols):
-                    subframe = subframe[subframe[gcol] == group[idx]]
+            folds = grouped_ts_splitter(data, k, gcols)
```
Comment (hakunanatasha) [inline lightwood/data/splitter.py:25]: happy about this abstraction, immensely.

File: lightwood/data/splitter.py [hunk 3]
```diff
@@ -19,16 +22,27 @@ def splitter(data: pd.DataFrame, k: int, tss: TimeseriesSettings) -> List[pd.Dat
...
+    return folds
+
+
+def grouped_ts_splitter(data: pd.DataFrame, k: int, gcols: List[str]):
+    """
+    Splitter for grouped time series tasks, where there is a set of `gcols` columns by which data is grouped.
+    Each group yields a different time series, and the splitter generates `k` folds from `data`,
+    with equally-sized sub-series for each group.
+    """
+    all_group_combinations = list(product(*[data[gcol].unique() for gcol in gcols]))
```
Comment (hakunanatasha) [inline lightwood/data/splitter.py:36]: I'm actually not fully aware what product(*<var>) does. Can you elaborate?

File: lightwood/data/splitter.py [hunk 4]
```diff
@@ -19,16 +22,27 @@ def splitter(data: pd.DataFrame, k: int, tss: TimeseriesSettings) -> List[pd.Dat
...
+    folds = [pd.DataFrame() for _ in range(k)]
```
Comment (hakunanatasha) [inline lightwood/data/splitter.py:37]: please explicitly define folds, alternatively, call them subsets (as I believe it's closer to the intended goal). I recognize we will likely change splitter's intended role, but for the sake of clarity it helps.

File: lightwood/data/splitter.py [hunk 5]
```diff
@@ -19,16 +22,27 @@ def splitter(data: pd.DataFrame, k: int, tss: TimeseriesSettings) -> List[pd.Dat
...
+    for group in all_group_combinations:
+        subframe = data
```
Comment (hakunanatasha) [inline lightwood/data/splitter.py:39]: This might not apply: do we want to use deepcopy here as opposed to subframe=data?

**Review by hakunanatasha — CHANGES_REQUESTED** [review]

**Review by hakunanatasha — CHANGES_REQUESTED** [review]
As an aside, why wouldn't we add this to the helpers function folder?

File: lightwood/data/splitter.py (hunk 4 above)
Comment (paxcema) [reply lightwood/data/splitter.py:37]: Yes, agreed on renaming to `subsets`

File: lightwood/data/splitter.py (hunk 1 above)
Comment (paxcema) [reply lightwood/data/splitter.py:11]: Indeed, equally sized. As discussed elsewhere, will be renaming to `subset`.

File: lightwood/data/splitter.py (hunk 3 above)
Comment (paxcema) [reply lightwood/data/splitter.py:36]: It yields the cartesian product of all observed values for each `group_by` column.

So, if `gby_col_a` has `A, B` as possible values, and `gby_col_b` has `C, D` as possible values, this line of code determines that all possible time series are given by the following combinations:

`[(A, C), (B, C), (A, D), (B, D)]`

File: lightwood/data/splitter.py (hunk 5 above)
Comment (paxcema) [reply lightwood/data/splitter.py:39]: Well, these subsets should form a partition over the original dataframe, so I'd think it's not a problem.

**Comment by paxcema** [comment]
> As an aside, why wouldn't we add this to the helpers function folder?

We could, though I'm partial to leaving all splitting logic here, else this module would end up empty. I defer the choice to you, though.

File: lightwood/data/splitter.py (hunk 3 above)
Comment (hakunanatasha) [reply lightwood/data/splitter.py:36]: Great - be nice to make a small note of that in the docstring explicitly with the phrase you used earlier ("It yields the cartesian product of all observed values for each group_by column.")

**Review by hakunanatasha — APPROVED** [review]

**Pull request merged**